### PR TITLE
Fixes #6: Rewrite turn functions to be pure

### LINKED ISSUE #6: rewrite turn functions to be pure
right now the four turn functions in `Game.js` (`playCard`, `aboutToPlay`, `cancelPlay`, and `resetGame`) all mutate the game state `G`. 

Rewrite them to treat `G` as immutable.

### PR DESCRIPTION
**Pull request by rezziemaven**
Changes made re: issue #6:
- Made playCard() immutable
- Made aboutToPlay() immutable
- Made cancelPlay() immutable

Don't think resetGame() needed modifying as it wasn't returning a modified version of G but rather the init object from setupGame().

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -24,9 +24,9 @@ import { Lobby } from 'boardgame.io/react';
 //   )
 // }
 
-const server = `https://${window.location.hostname}`;
+// const server = `https://${window.location.hostname}`;
 // development mode
-// const server = http://${window.location.hostname}:8000
+const server =  `http://${window.location.hostname}:8000`;
```
Comment (MustafaHaddara) [inline src/App.js]: can you revert this change? that way the code is good to be deployed as-is.

File: src/Game.js [hunk 2]
```diff
@@ -74,25 +74,42 @@ const playCard = (G, ctx, card) => {
...
+    return {
+        ...G,
+        players: {
+            [playerId]: {
+                cards: G.players[playerId].cards.filter(c => c !== playedCard)
+            }
+        },
+        aboutToPlay: {
+            [playerId]: null
+        },
```
Comment (MustafaHaddara) [inline src/Game.js]: same as above, this deletes the entire contents `aboutToPlay`...can you change it to something like this?
```suggestion
        aboutToPlay: {
            ...aboutToPlay
            [playerId]: null
        },
```

File: src/Game.js [hunk 3]
```diff
@@ -74,25 +74,42 @@ const playCard = (G, ctx, card) => {
...
-    G.aboutToPlay[playerId] = null;
-
-    G.cardsPlayed.push(playedCard);
-    G.gameOver = isGameOver(G);
+    return {
+        ...G,
+        players: {
+            [playerId]: {
+                cards: G.players[playerId].cards.filter(c => c !== playedCard)
+            }
```
Comment (MustafaHaddara) [inline src/Game.js:83]: this overwrites the entire contents of `players`...can you change it to something like this? 
```
            ...players
            [playerId]: {
                cards: G.players[playerId].cards.filter(c => c !== playedCard)
            }
```

File: src/Game.js [hunk 4]
```diff
@@ -74,25 +74,42 @@ const playCard = (G, ctx, card) => {
...
+        },
+        aboutToPlay: {
+            [playerId]: null
+        },
+        cardsPlayed: [...G.cardsPlayed, playedCard],
+        gameOver: isGameOver(G)
```
Comment (MustafaHaddara) [inline src/Game.js]: oh damn, this might not work. :/ 

I'm pretty sure `isGameOver` needs the _updated_ `G`.

File: src/Game.js [hunk 5]
```diff
@@ -74,25 +74,42 @@ const playCard = (G, ctx, card) => {
...
+    }
 };
 
 const aboutToPlay = (G, ctx) => {
-    G.aboutToPlay[ctx.playerID] = Date.now();
+    return {
+        ...G,
+        aboutToPlay: {
+            [ctx.playerID]: Date.now()
+        }
```
Comment (MustafaHaddara) [inline src/Game.js]: same as above, this deletes the entire contents `aboutToPlay`...can you change it to something like this?
```suggestion
        aboutToPlay: {
            ...aboutToPlay
            [ctx.playerId]: Date.now()
        },
```

File: src/Game.js [hunk 6]
```diff
@@ -74,25 +74,42 @@ const playCard = (G, ctx, card) => {
...
+    }
 };
 
 const cancelPlay = (G, ctx) => {
-    G.aboutToPlay[ctx.playerID] = null;
+    return {
+        ...G,
+        aboutToPlay: {
+            [ctx.playerID]: null
+        }
```
Comment (MustafaHaddara) [inline src/Game.js]: same as above, this deletes the entire contents `aboutToPlay`...can you change it to something like this?
```suggestion
        aboutToPlay: {
            ...aboutToPlay
            [playerId]: null
        },
```

**Review by MustafaHaddara — CHANGES_REQUESTED** [review]
Looks like a solid start, I left some comments.

Thanks!

File: src/App.js (hunk 1 above)
Comment (rezziemaven) [reply src/App.js]: Ah, sorry about that! Just fixed in latest commits.

File: src/Game.js (hunk 2 above)
Comment (rezziemaven) [reply src/Game.js]: Oh shoot, my bad! Just fixed.

File: src/Game.js (hunk 3 above)
Comment (rezziemaven) [reply src/Game.js:83]: Sorry! Just fixed.

File: src/Game.js (hunk 4 above)
Comment (rezziemaven) [reply src/Game.js]: Ah, so I think I may have figured out a way to still get it to work. Instead of returning the new object immediately, I decided to make a new object called newG that will act as the deep copy of G and then modifying newG.gameOver after that object was created. I hope that should solve the problem!

File: src/Game.js (hunk 5 above)
Comment (rezziemaven) [reply src/Game.js]: Np! Fixed.

File: src/Game.js (hunk 6 above)
Comment (rezziemaven) [reply src/Game.js]: Gotcha! Fixed.

File: src/Game.js [hunk 7]
```diff
@@ -74,25 +74,29 @@ const playCard = (G, ctx, card) => {
...
         },
         aboutToPlay: {
+            ...G.aboutToPlay,
             [playerId]: null
         },
         cardsPlayed: [...G.cardsPlayed, playedCard],
-        gameOver: isGameOver(G)
     }
+    newG.gameOver = isGameOver(newG);
+    return newG;
```
Comment (MustafaHaddara) [inline src/Game.js:92]: 👍

**Review by MustafaHaddara — APPROVED** [review]
LGTM, thanks!
